addons/test_mail/tests/test_mail_resend.py

```diff
@@ -0,0 +1,180 @@
+        self.assertEqual(self.send_email_counter, 0)
+        message = self.test_record.sudo().message_post(partner_ids=self.partners.ids, subtype='mail.mt_comment', message_type='notification')
+        self.assertEqual(self.send_email_counter, 4)
+        self.assertEqual(self.success_email_counter, 2)
+        self.assertEqual(len(self.bus_update_failure), 2, 'two failure sent on bus, one for each mail')
+        wizard = self.env['mail.resend.message'].with_context({'mail_message_to_resend': message.id}).create({})
+        partners = wizard.partner_ids.mapped("partner_id")
+        self.assertEqual(self.invalid_email_partners, partners)
+        wizard.partner_ids.filtered(lambda p: p.partner_id == self.partner1).write({"is_to_send": False})
+        wizard.resend_mail_action()
+        self.assertEqual(self.send_email_counter, 5)
+        self.assertEqual(self.success_email_counter, 2)
+        self.assertEqual(len(self.bus_update_failure), 4, 'two more failure sent on bus, one for failure, one for resnd')
+        cancel = self.env['mail.notification'].search([('mail_message_id', '=', message.id), ('res_partner_id', '=', self.partner1.id)])
+        other = self.env['mail.notification'].search([('mail_message_id', '=', message.id), ('res_partner_id', '=', self.user1.partner_id.id)])
+        self.assertEqual(('exception', 'sent', 'canceled', 'sent'), tuple(self._notif_by_partner(message).mapped('email_status')))
+        self.user1.write({"email": 'u1@example.com'})
+        wizard = self.env['mail.resend.message'].with_context({'mail_message_to_resend': message.id}).create({})
+        self.assertEqual(self.user1.partner_id, wizard.partner_ids.mapped("partner_id"))
+        wizard.resend_mail_action()
+        self.assertEqual(self.send_email_counter, 6)
+        self.assertEqual(self.success_email_counter, 3)
+        self.assertEquals(('sent', 'sent', 'canceled', 'sent'), tuple(self._notif_by_partner(message).mapped('email_status')))
+        self.assertEqual(len(self.bus_update_failure), 5, "No more notification in failure, a notification should have been sent by resend")
+
+    @mute_logger('odoo.addons.mail.models.mail_mail')
+    def test_cancel_all(self):
+        message = self.test_record.sudo().message_post(partner_ids=self.partners.ids, subtype='mail.mt_comment', message_type='notification')
+        self.assertEqual(('exception', 'sent', 'exception', 'sent'), tuple(self._notif_by_partner(message).mapped('email_status')))
+        wizard = self.env['mail.resend.message'].with_context({'mail_message_to_resend': message.id}).create({})
+        wizard.cancel_mail_action()
+        self.assertEqual(('canceled', 'sent', 'canceled', 'sent'), tuple(self._notif_by_partner(message).mapped('email_status')))
+
+    @mute_logger('odoo.addons.mail.models.mail_mail')
+    def test_bounce_resend(self):
+        return
+
+    @mute_logger('odoo.addons.mail.models.mail_mail')
+    def test_layout_resend(self):

```

Comment:
I don't really understand the purpose of that test.

Reply:
To remove before merge